### PR TITLE
Templating grammar & visitor refactor

### PR DESCRIPTION
**Pull request by alexandraorth**
+ Improved the templating grammar 
+ Refactored code in the template visitor
+ Added array accessors for lists in templates (i.e. list[0])

### CODE REVIEW

**Comment by alexandraorth** [comment]
!rtg - cassandra timeout

**Comment by alexandraorth** [comment]
@aelred can you review this one, because templating

File: grakn-graql/src/main/java/ai/grakn/graql/internal/template/TemplateVisitor.java [hunk 1]
```diff
@@ -151,98 +122,121 @@ public String visitIfStatement(GraqlTemplateParser.IfStatementContext ctx){
...
     @Override
     public Boolean visitEqExpression(GraqlTemplateParser.EqExpressionContext ctx) {
-        Object lValue = this.visit(ctx.expr(0));
-        Object rValue = this.visit(ctx.expr(1));
+        Object lValue = this.visit(ctx.expression(0));
+        Object rValue = this.visit(ctx.expression(1));
+
+        if(lValue == null || rValue == null){
+            return lValue == rValue;
+        }
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/template/TemplateVisitor.java]: bit weird

File: grakn-graql/src/main/java/ai/grakn/graql/internal/template/TemplateVisitor.java [hunk 2]
```diff
@@ -379,49 +311,87 @@ public String visitTerminal(TerminalNode node){
...
+        return StringUtil.valueToString(resolved);
     }
 
-    private Object concat(Object... values){
-        if(values.length == 1){
-            return values[0];
-        }
+    @Override
+    public Object visitMacroExpression(GraqlTemplateParser.MacroExpressionContext ctx){
+        String macro = ctx.ID_MACRO().getText().replace("@", "").toLowerCase();
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/template/TemplateVisitor.java]: We should probably just use the substring method here.

**Pull request merged**